File: tablix/public/js/controllers/manual_assign.js

/*
 * 	Added By Sahil
 * 	
 */
frappe.provide("tablix.manual_assign")

tablix.manual_assign.ManualAssign = Class.extend({
	make_dialog: function(){

		var me = this;
		var user_info = frappe.user_info();
		var user = user_info.name;
		var title = frappe.utils.format("Assign {0} - {1}", [me.frm.doc.doctype, me.frm.doc.name]);
		this.dialog = new frappe.ui.Dialog({
			title: __(title),
			fields:[
				{reqd: 1, fieldname: "assign_to", fieldtype:"Link", options: "User",label:"Assign To"},
				{reqd: 1, fieldname: "assign_by", fieldtype: "Link", options:"User", default:user, label:"Assign By"},
				{reqd: 1, fieldname: "comment", fieldtype:"Small Text", label:"Add Comment",},
				{fieldname: "change_status", fieldtype: "Check", label: "Change Status"},
				{fieldname: "status", fieldtype: "Link", options: "Workflow State"},
				{fieldname: "is_notify", fieldtype: "Check", label: __("Notify Employee"), default:1},
				{fieldname: "doctype", fieldtype: "Data", read_only:1, default:me.frm.doctype}
			],
			primary_action_label: __("Assign"),
			primary_action: function(frm){
				var doc = me.frm.doc;
				if(me.frm.is_dirty() || me.frm.is_new()){
					frappe.msgprint("Please save the change and try again");
					return false;
				}
				me.validate_mandatory(frm);
				me.dialog.hide();
				$.extend(frm, {frm:doc});
				frm.status_field = doc.tablix_status?"tablix_status": null
				frappe.call({
					method: "tablix.whitelisted.manual_assign",
					args: frm,
					callback: function(res){
						if(res && res.message){
							frappe.msgprint("Assignment Successfull");
							me.make_comment(frm);
							me.frm.refresh();
						}
						else{
							frappe.msgprint("Something went wrong");
							return;
						}
					}
				});
			}
		});
		this.frm.add_custom_button(__("Manual Assignment"), function(){
	
			if(me.frm.doctype==="Task"){
				frappe.msgprint(__("You can't assign <b>Task</b> Manually, Please use <b>Action Button</b>"));
				return false;
			}	
			if(me.frm.is_dirty()===1){
				frappe.msgprint(__("Please save the document before <b>Manual Assignment </b>"));
				return false;
			}
			me.dialog.show();

		}).addClass("btn-secondary");	
	},
	make_comment: function(args){
		var msg  = frappe.utils.format("Assign By {0}, Assign to {1} for {2}", [1, args.assign_by, args.assign_to, args.comment]);
		this.frm.timeline.insert_comment("Assigned",  msg);
	},
	validate_mandatory: function(args){

		if(args.change_status && ! args.status){
			frappe.msgprint(_("Please select the Valid Status"));
			return false;
		}
	},

	get_status: function(){
		var doctype  = this.frm.doctype;
		var docname = this.frm.docname;
		var states = [];
		if(doctype == "Boq"){
		}
	},
});

File: tablix/notifications/manual_assign.py

'''
	Developer Sahil
	Email sahil.saini@tablix.ae
    
'''

import frappe
from frappe import _, msgprint, throw
from frappe.utils import cint, flt, cstr
from .notifications import notify_employee
import json

def manual_assign(clr=True, *args, **kwargs):
        
        if not kwargs.get("comment") or not kwargs.get("assign_to") or not kwargs.get("assign_by"):
                throw(_("Mandatory Fields <b> Comment, Assign To, Assign By</b>"))

        frm = kwargs.get("frm")
        if isinstance(frm, str):
                frm = json.loads(frm)

        doc = frappe.get_doc(frm.get("doctype"), frm.get("name"))
        if doc.docstatus == 2:
                throw(_("You cannot assign <b>Cancelled</b> document. \n Please <b>Cancel and Amend</b> \
                        document then assign the document"))
        
        try:
                update_status(doc, **kwargs)
                send_email(doc, **kwargs)
                assign_to(doc, clr=clr, **kwargs)
        except Exception as e:
                print(frappe.get_traceback())
                return {"message": "Failure"}

        return {"message": "Success"}


def update_status(doc, **kwargs):
	setting  = frappe.get_doc("Tablix Setting", "Tablix Setting")
	if kwargs.get("status") and kwargs.get("status_field"):
		if not doc.get("tablix_status"):
			return
		elif doc.get("tablix_status") == kwargs.get("status"):
			throw(_("Please select different status"))

		if not setting.get("allow_change_status"):
			frappe.msgprint(_("You can't change the <b>Status</b> but you can still assign"))
			return 
		frappe.db.set_value(doc.get("doctype"), doc.get("name"), "tablix_status", kwargs.get("status"))

def send_email(doc, **kwargs):
	if not kwargs.get("is_notify") or kwargs.get("is_notify") == '0':
		return
	emails = [kwargs.get("assign_to")]
	msg = kwargs.get("comment")
	notify_employee(doc, emails, msg)

def assign_to(doc, clr=True, **kwargs):
	from frappe.desk.form.assign_to import add, clear
	if clr:
		clear(doc.get("doctype"), doc.get("name"))

	add({
		"assign_to": kwargs.get("assign_to"),
		"name": doc.get("name"),
		"doctype": doc.get("doctype"),
		"description": kwargs.get("comment")
	})
